### ISSUE TITLE
12 hour format datetimes with meridian not being patched correctly

### ISSUE THREAD

**Issue by JayWalker512** [cakephp/cakephp#7430]
Howdy, I'm having an issue with an application I'm developing where sometimes dates passed in 12-hour format with am/pm will not be processed correctly by patchEntity() method. 

For context my app has a form for employees to fill out time sheets, and some valid entries will not be patched such as a start time of 10:30am and end time of 12:30pm. When these values are entered, only the start time of 10:30am will exist in the patched entity. Both fields have the same validation rules. I'll supply everything below such that you should be able to reproduce the issue, let me know if you need anything more.

Here's my 'add' form
```php
<?= $this->Form->create($timeSheet) ?>
        <fieldset>
            <legend><?= __('Add Time Sheet') ?></legend>
            <?php
            if ($logged_user->is_admin) { //check to see if user is admin
                echo $this->Form->input('sp_id', [
                    'type' => 'select',
                    'options' => $sps,
                    'default' => $current_sp_id,
                    'label' => 'Select an ID other than your own if you need to create a time sheet for another SP.'
                ]);
            }

            echo $this->Form->input('start_time', ['type' => 'datetime', 
                                                   'interval' => 5,
                                                   'minYear' => date("Y") - 1,
                                                   'maxYear' => date("Y"),
                                                   'timeFormat' => 12,
                                                   'label' => "Start Date and Time"]);
            echo $this->Form->input('end_time', ['type' => 'datetime', 
                                                   'interval' => 5,
                                                   'minYear' => date("Y") - 1,
                                                   'maxYear' => date("Y"),
                                                   'timeFormat' => 12,
                                                   'label' => "End Date and Time"]);
            echo $this->Form->input('program', ['type' => 'integer',
                                                'label' => 'Payroll Code']);
            echo $this->Form->input('sp_role_id', ['options' => $sp_roles, 'empty' => false]);
            echo $this->Form->input('notes', ['type' => 'textarea',
                                              'label' => 'Notes (optional)']);
            ?>
        </fieldset>
        <?= $this->Form->button(__('Submit')) ?>
        <?= $this->Form->end() ?>
```

Controller code I'm currently using to display that the data isn't being patched correctly
```php
public function add()
{   
        $timeSheet = $this->TimeSheets->newEntity();
        if ($this->request->is('post')) {    
            
            $timeSheet = $this->TimeSheets->patchEntity($timeSheet, $this->request->data);
            
            echo "<pre>Request data:";
            echo print_r($this->request->data);
            echo "Patched data:";
            echo print_r($timeSheet);
            echo "</pre>";

//code below here is to popular other drop down fields in the add form (omitted)

}
```

Validation rules
```php
        $validator
            ->add('start_time', 'valid', ['rule' => 'datetime'])
            ->requirePresence('start_time', 'create')
            ->notEmpty('start_time');
            
        $validator
            ->add('end_time', 'valid', ['rule' => 'datetime'])
            ->requirePresence('end_time', 'create')
            ->notEmpty('end_time');
```

Request data submitted by form:
```html
Request data:Array
(
    [sp_id] => 520
    [start_time] => Array
        (
            [year] => 2015
            [month] => 09
            [day] => 21
            [hour] => 10
            [minute] => 30
            [meridian] => am
        )

    [end_time] => Array
        (
            [year] => 2015
            [month] => 09
            [day] => 21
            [hour] => 12
            [minute] => 30
            [meridian] => pm
        )

    [program] => 22
    [sp_role_id] => 1
    [notes] => 
)
```

And the data present in the patched entity. You can see here that only 'start_time' exists
```html
Patched data:App\Model\Entity\TimeSheet Object
(
    [_accessible:protected] => Array
        (
            [sp_id] => 1
            [start_time] => 1
            [end_time] => 1
            [week_ending_date] => 1
            [date_submitted] => 1
            [program] => 1
            [rate] => 1
            [benefits] => 1
            [total_cost] => 1
            [notes] => 1
            [sp_role_id] => 1
            [sp] => 1
        )

    [_properties:protected] => Array
        (
            [sp_id] => 520
            [start_time] => Cake\I18n\Time Object
                (
                    [date] => 2015-09-21 10:30:00
                    [timezone_type] => 3
                    [timezone] => America/New_York
                )

            [program] => 22
            [sp_role_id] => 1
            [notes] => 
        )

    [_original:protected] => Array
        (
        )

    [_hidden:protected] => Array
        (
        )

    [_virtual:protected] => Array
        (
        )

    [_className:protected] => App\Model\Entity\TimeSheet
    [_dirty:protected] => Array
        (
            [sp_id] => 1
            [start_time] => 1
            [program] => 1
            [sp_role_id] => 1
            [notes] => 1
        )

    [_new:protected] => 1
    [_errors:protected] => Array
        (
            [end_time] => Array
                (
                    [valid] => The provided value is invalid
                )

            [week_ending_date] => Array
                (
                    [_required] => This field is required
                )

            [date_submitted] => Array
                (
                    [_required] => This field is required
                )

            [rate] => Array
                (
                    [_required] => This field is required
                )

            [benefits] => Array
                (
                    [_required] => This field is required
                )

            [total_cost] => Array
                (
                    [_required] => This field is required
                )

        )

    [_registryAlias:protected] => TimeSheets
)
```

Version of Cake as displayed in /vendor/cakephp/cakephp/VERSION.txt is 3.0.13

**Comment by markstory** [comment]
Your entity, has the `end_date` marked as invalid, which is why the data has not been added to the entity. Oddly, the following test case passes:

```php
$date = ['year' => 2015, 'month' => 9, 'day' => 21, 'hour' => 12, 'minute' => 30, 'meridian' => 'pm'];
$this->assertTrue(Validation::datetime($date));
```

Are there any other validation rules on the end_date?

**Comment by JayWalker512** [comment]
Not that I'm aware of, the validation rules I pasted in are the only ones that should have any effect.

**Comment by markstory** [comment]
Ok. I'll try reproducing the issue in the context of an application as the data you've provided is considered 'valid' as far as a unit test is concerned.

**Comment by markstory** [comment]
I tried setting up something similar in a local testing application and the end_time field was marshalled correctly. Also in your original code sample the `start_time` field has a similiar input definition, and validation rule however that input was marshalled correctly.

Given that you have one working field with a 12 hour format, I recommend you figure out what is different between start_time and end_time and go from there.

**Comment by JayWalker512** [comment]
I'm not sure why you weren't able to reproduce the issue, but I've reproduced it myself using a barebones  fresh project. I've included instructions below to reproduce the issue, as well as the results of my investigation of the bug:

Steps to reproduce
===============

1. Create database we'll use in our CakePHP project using script below:
-------------------------------------------------------------------------------------------------
```sql
DROP DATABASE meridian_db;
CREATE DATABASE meridian_db;
USE meridian_db;

CREATE TABLE time_sheets (
	id INT UNSIGNED AUTO_INCREMENT PRIMARY KEY,
	start_time DATETIME NOT NULL,
	end_time DATETIME NOT NULL,
	rate INT UNSIGNED NOT NULL,
	notes TEXT
);
```
2. Use composer to create the CakePHP project
-------------------------------------------------------------------------------------------------
Command is: php composer.phar create-project --prefer-dist cakephp/app [app_name]

3. Bake all files necessary to have a basic CRUD application
-------------------------------------------------------------------------------------------------
Command is: bin/cake bake all time_sheets

4. Insert proper DB credentials in [app_name]/config/app.php 
-------------------------------------------------------------------------------------------------
5. Change add/edit view input fields to 12 hour format. Something like this:
-------------------------------------------------------------------------------------------------
```php
echo $this->Form->input('start_time', ['type' => 'datetime', 
   'interval' => 5,
   'minYear' => date("Y") - 1,
   'maxYear' => date("Y"),
   'timeFormat' => 12,
   'label' => "Start Date and Time"]);
echo $this->Form->input('end_time', ['type' => 'datetime', 
   'interval' => 5,
   'minYear' => date("Y") - 1,
   'maxYear' => date("Y"),
   'timeFormat' => 12,
   'label' => "End Date and Time"]);
```
6. Navigate a browser to [yourprojecturl]/time_sheets/add
-------------------------------------------------------------------------------------------------
7. Populate start and end time fields with todays date, and 12:10pm and 12:50pm, respectively. Populate other fields however you like.
-------------------------------------------------------------------------------------------------
8. Click submit, and observe that the times are both marked as invalid.
-------------------------------------------------------------------------------------------------

My investigation results
==================

So I spent some time debugging and have discovered the source of the bug, which seems to be in CakePHP core (specifically, Validation.php). 

The reason your test case passes is because the values in the data array containing time information are stored as strings, and not ints. See this screenshot:

![screenshot from 2015-09-24 16 32 54](https://cloud.githubusercontent.com/assets/1485718/10151511/49d00244-6616-11e5-826d-65bbf282c2be.png)


And due to the way the validation check works in Validation.php, arrays of ints will pass where arrays of strings will not. See for yourself with the tests below, the first will pass and the latter will fail:

```php
$date = ['year' => 2015, 'month' => 9, 'day' => 21, 'hour' => 12, 'minute' => 30, 'meridian' => 'pm'];
        $this->assertTrue(Validation::datetime($date));
        
$date = ['year' => '2015', 'month' => '09', 'day' => '21', 'hour' => '12', 'minute' => '30', 'meridian' => 'pm'];
        $this->assertTrue(Validation::datetime($date));
```

I've narrowed the problem down to a strange bit of code in Validation.php (around line 1184 in the version I cloned using the reproduction instructions above, the function in question is '_getDateString()') that will convert a string type hour equalling 12 to 24, which is then marked as an invalid time by a regular expression comparison farther down the line in the validation process. I did all this using xdebug, which was very handy. 

I tried fixing the bug myself, but ended up with more problems than I started with. So hopefully the information I've provided here can be helpful to you guys.

**Comment by markstory** [comment]
Thanks for the detailed steps @JayWalker512 I will tske another look.

**Comment by markstory** [comment]
@JayWalker512 I think the changes in #7476 fix the issue - at least they did for me.

**Comment by JayWalker512** [comment]
With those changes the input is validated, but it is saved incorrectly. 12:05pm and 12:55pm are not the same as 00:05 and 00:55, but they are saved as such (in the next calendar day, no less):

```php
Array
        (
            [sp_id] => 520
            [start_time] => Cake\I18n\Time Object
                (
                    [date] => 2015-09-30 00:05:00
                    [timezone_type] => 3
                    [timezone] => America/New_York
                )

            [end_time] => Cake\I18n\Time Object
                (
                    [date] => 2015-09-30 00:55:00
                    [timezone_type] => 3
                    [timezone] => America/New_York
                )

            [program] => 9
            [sp_role_id] => 1
            [notes] => 
        )
```

![incorrect-time-saved](https://cloud.githubusercontent.com/assets/1485718/10168595/7dd13ca4-669b-11e5-9c95-39c5f8ba3a3c.png)

**Comment by markstory** [comment]
Ok, in addition to the validation code having a problem, the datetime marshalling code also had an issue. Both issues should be fixed when #7485 merges in. Thanks for putting up with me being dense.
